x-pack/plugins/code/server/ctags/ctags_reader.ts

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Tag, TagFields } from '../../model';
+import { LoggerFactory } from '../utils/log_factory';
+
+export class CtagsReader {
+  private tags: Tag[] = [];
+
+  private MIN_METHOD_LINE_LENGTH = 6;
+  private MAX_METHOD_LINE_LENGTH = 1030;
+
+  constructor(
+    readonly loggerFactory: LoggerFactory
+  ) {}
+
+  public getTags(): Tag[] {
+    return this.tags;
+  }
+
+  public readLine(tagLine: string) {
+    // const logger = this.loggerFactory.getLogger(['code', 'ctagsReader']);
+
+    if (tagLine === null) {
+      return;
+    }
+
+    let p = tagLine.indexOf('\t');
+    if (p <= 0) {
+      return;
+    }
+    const def = tagLine.substring(0, p);
+    const mstart = tagLine.indexOf('\t', p + 1);
+
+    let kind = null;
+
+    let lp = tagLine.length;
+    let tagFieldMap = new Map();
+    while((p = tagLine.lastIndexOf('\t', lp - 1)) > 0) {
+      const fld = tagLine.substring(p + 1, lp);
+      lp = p;
+
+      const sep = fld.indexOf(':');
+      let matchTagField = null;
+      if (sep != -1) {
+        for (let tagField in TagFields) {
+          if (TagFields[tagField].charAt(0) !== fld.charAt(0)) {
+            continue;
+          } else {
+            matchTagField = TagFields[tagField];
+          }
+        }
+        if (matchTagField !== null) {
+          tagFieldMap.set(matchTagField, fld.substring(sep + 1));
+        }
+      } else {
+          kind = fld;
+          break;
+      }
+    }
+
+    let lnum = tagFieldMap.get(TagFields.LINE);
+    let signature = tagFieldMap.get(TagFields.SIGNATURE);
+    let classInher = tagFieldMap.get(TagFields.CLASS);
+
+    let whole;
+    let match;
+    const mlength = p - mstart;
+    if ((p > 0) && (mlength > this.MIN_METHOD_LINE_LENGTH)) {
+      whole = this.cutPattern(tagLine, mstart, p);
+      if (mlength < this.MAX_METHOD_LINE_LENGTH) {
+        match = whole.replace(/[ \t]+/g, ' ');
+      } else {
+        match = whole.substring(0, this.MAX_METHOD_LINE_LENGTH).replace(/[ \t]+/g, ' ');
+      }
+    } else {
+      return;
```

Comment:
Another coding style comment! We can use early return in the `if` stmt to avoid `else{ return; }`. @mw-ding ?
```ts
if ((p <= 0) || (mlength <= this.MIN_METHOD_LINE_LENGTH)) {// Or put ! at the front of condition.
  return;
}

whole = this.cutPattern(tagLine, mstart, p);
if (mlength < this.MAX_METHOD_LINE_LENGTH) {
   match = whole.replace(/[ \t]+/g, ' ');
} else {
    match = whole.substring(0, this.MAX_METHOD_LINE_LENGTH).replace(/[ \t]+/g, ' ');
}
```

Reply:
Frankly, I don't have a strong opinion on this as long as it's not crazily confusing and unreadable. To me, I would suggest providing more meaningful variable names to replace `p` and `mlength` (what is `m` in this case?)